File: python/open_guardrail/guards/language_zh.py
"""Detect Chinese language presence."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_CHINESE = re.compile(
    r"[\u4E00-\u9FFF\u3400-\u4DBF\uF900-\uFAFF]"
)


class _LanguageZh:
    def __init__(
        self, *, action: str = "warn", min_ratio: float = 0.3
    ) -> None:
        self.name = "language-zh"
        self.action = action
        self._min_ratio = min_ratio

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        stripped = re.sub(r"\s", "", text)
        if not stripped:
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="language-zh",
                passed=False,
                action=self.action,
                score=0.0,
                latency_ms=round(elapsed, 2),
            )

        matches = _CHINESE.findall(stripped)
        ratio = len(matches) / len(stripped)
        passed = ratio >= self._min_ratio
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="language-zh",
            passed=passed,
            action="allow" if passed else self.action,
            score=round(ratio, 3),
            latency_ms=round(elapsed, 2),
            details={"chinese_ratio": round(ratio, 3)},
        )


def language_zh(
    *, action: str = "warn", min_ratio: float = 0.3
) -> _LanguageZh:
    return _LanguageZh(action=action, min_ratio=min_ratio)

File: packages/guards/src/language-zh.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface LanguageZhOptions {
  action: 'block' | 'warn';
  minRatio?: number;
}

const CHINESE_RANGE = /[\u4E00-\u9FFF\u3400-\u4DBF\uF900-\uFAFF]/g;

export function languageZh(options: LanguageZhOptions): Guard {
  const minRatio = options.minRatio ?? 0.3;

  return {
    name: 'language-zh',
    version: '0.1.0',
    description: 'Detects Chinese language presence',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const stripped = text.replace(/\s/g, '');
      if (stripped.length === 0) {
        return {
          guardName: 'language-zh',
          passed: false,
          action: options.action,
          score: 0,
          latencyMs: Math.round(performance.now() - start),
        };
      }

      const matches = stripped.match(CHINESE_RANGE) || [];
      const ratio = matches.length / stripped.length;
      const passed = ratio >= minRatio;

      return {
        guardName: 'language-zh',
        passed,
        action: passed ? 'allow' : options.action,
        score: Math.round(ratio * 1000) / 1000,
        latencyMs: Math.round(performance.now() - start),
        details: { chineseRatio: Math.round(ratio * 1000) / 1000 },
      };
    },
  };
}
